pkg/kubelet/cm/container_manager_linux_test.go

```diff
@@ -268,3 +268,69 @@ func TestGetCapacity(t *testing.T) {
 		})
 	}
 }
+
+func TestNewPodContainerManager(t *testing.T) {
+	cases := []struct {
+		name string
+		cm   *containerManagerImpl
+	}{
+		{
+			name: "CgroupsPerQOS is disabled, return *podContainerManagerNoop",
+			cm: &containerManagerImpl{
+				qosContainerManager: &qosContainerManagerImpl{
+					qosContainersInfo: QOSContainersInfo{
+						Guaranteed: CgroupName{"guaranteed"},
+						BestEffort: CgroupName{"besteffort"},
+						Burstable:  CgroupName{"burstable"},
+					},
+					subsystems:         &CgroupSubsystems{},
+					cgroupManager:      NewCgroupManager(&CgroupSubsystems{}, ""),
+					activePods:         func() []*v1.Pod { return []*v1.Pod{} },
+					getNodeAllocatable: func() v1.ResourceList { return v1.ResourceList{} },
+					cgroupRoot:         CgroupName{"burstable"},
+					qosReserved:        map[v1.ResourceName]int64{},
+				},
+
+				NodeConfig: NodeConfig{
+					CgroupsPerQOS: false,
+				},
+			},
+		},
+		{
+			name: "CgroupsPerQOS is enabled, return *podContainerManagerImpl",
+			cm: &containerManagerImpl{
+				qosContainerManager: &qosContainerManagerImpl{
+					qosContainersInfo: QOSContainersInfo{
+						Guaranteed: CgroupName{"guaranteed"},
+						BestEffort: CgroupName{"besteffort"},
+						Burstable:  CgroupName{"burstable"},
+					},
+					subsystems:         &CgroupSubsystems{},
+					cgroupManager:      NewCgroupManager(&CgroupSubsystems{}, ""),
+					activePods:         func() []*v1.Pod { return []*v1.Pod{} },
+					getNodeAllocatable: func() v1.ResourceList { return v1.ResourceList{} },
+					cgroupRoot:         CgroupName{"burstable"},
+					qosReserved:        map[v1.ResourceName]int64{},
+				},
+
+				NodeConfig: NodeConfig{
+					CgroupsPerQOS: true,
+				},
+			},
+		},
+	}
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			pcm := c.cm.NewPodContainerManager()
+			if c.cm.NodeConfig.CgroupsPerQOS {
+				if _, ok := pcm.(*podContainerManagerImpl); !ok {
+					t.Errorf("got unexpected type %T, expecting *podContainerManagerImpl", pcm)
```

Comment:
I think we could put the type here instead of PodContainerManager object:
```
t.Errorf("got unexpected type %T, expecting *podContainerManagerImpl", reflect.TypeOf(pcm))

```

Reply:
Actually, if we wanted to use reflect.TypeOf(pcm). We could refactor like this instead:
```
pcm := c.cm.NewPodContainerManager()
pcmtype := reflect.TypeOf(pcm)

if c.cm.NodeConfig.CgroupsPerQOS {
	if type != *podContainerManagerImpl {
		t.Errorf("got unexpected type %s, expecting *podContainerManagerImpl", pcmtype)
	}
} else {
	if type != *podContainerManagerNoop {
		t.Errorf("got unexpected type %s, expecting *podContainerManagerNoop", pcmtype)
	}
}
```

But it really is preference and your way doesn't require the use of an additional import.

Feel free to disregard my comment :smile:

Reply:
Resolving based on our discussions